### PR TITLE
feat: Toolbar tool which enables to add Items into UI container to invoke different functionality in Reveal

### PR DESCRIPTION
**Pull request by pramodcog**

### CODE REVIEW

File: viewer/packages/tools/src/Toolbar/Toolbar.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import css from './Toolbar.css';
+
+/**
+ * Toolbar position within the canvas
+ */
+export enum ToolbarPosition {
+  Top = 'Top',
+  Bottom = 'Bottom',
+  Left = 'Left',
+  Right = 'Right'
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/Toolbar.ts:14]: use pascalCase, i.e. `Top = 'top'` etc

File: viewer/packages/tools/src/Toolbar/Toolbar.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  /**
+   * Set the position of the toolbar container
+   * @param position ToolbarPosition value such as Top, Bottom, Left, Right within the canvas
+   */
+  public setPosition(position: ToolbarPosition): void {
+    switch (position) {
+      case 'Top':
+        this._toolbarContainer.classList.remove(this._activeContainerPosition);
+        this._activeContainerPosition = Toolbar.classnames.top;
+        this._toolbarContainer.classList.add(Toolbar.classnames.top);
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/Toolbar.ts:117]: Create a utility function (its fine its just a local function within this function) that updates style so you avoid code duplication below.

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  /**
+   * Add a Icon button into the Toolbar container
+   * @param toolTip Tooltip for the Icon button
+   * @param backgroundImage background image to be placed onto icon
+   * @param isToggle Is the icon button used as toggle
+   * @param onClick Click event callback function which will be used to perform custom functionlity of the user
+   */
+  public addToolbarItem(toolTip: string, backgroundImage: string, isToggle: boolean, onClick: () => void): void {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:31]: `backgroundImage` - is it an URL? Rename to `backgrounImageUrl`

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+   * Set the position of the toolbar container
+   * @param position ToolbarPosition value such as Top, Bottom, Left, Right within the canvas
+   */
+  public setPosition(position: ToolbarPosition): void {
+    this._toolbar.setPosition(position);
+  }
+
+  public dispose(): void {
+    super.dispose();
+  }
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:45]: No need to override when you only call `super.dispose()`
```suggestion
```

File: viewer/packages/tools/src/Toolbar/Toolbar.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  /**
+   * Create & adds a button icons into the toolbar container
+   * @param toolTip Tooltip message to added for the icon
+   * @param backgroundImage Icon image to be displayed
+   * @param isToggle Is the icon button used as toggle
+   * @param onClick Click event action for the icon
+   */
+  public addToolbarItem(toolTip: string, backgroundImage: string, isToggle: boolean, onClick: () => void): void {
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply viewer/packages/tools/src/Toolbar/Toolbar.ts:84]: We are using image as `string` with base64 format not as url

File: viewer/webpack.config.js [hunk 6]
```diff
@@ -70,9 +71,13 @@ module.exports = env => {
...
         },
         {
-          test: /\.(png|svg|jpg|jpeg|gif)$/,
+          test: /\.(svg|jpg|jpeg|gif)$/,
           use: ['file-loader']
         },
+        {
+          test: /\.(png)$/,
+          use: ['url-loader']
+        },
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply viewer/webpack.config.js:80]: We are loading the PNG as base64 format string which will be loaded through `url-loader`

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts (hunk 3 above)
Comment (pramodcog) [reply viewer/packages/tools/src/Toolbar/ToolbarTool.ts:31]: We are loading the PNG as base64 format string which will be loaded through `url-loader`

File: viewer/packages/tools/src/Toolbar/Toolbar.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (larsmoa) [reply viewer/packages/tools/src/Toolbar/Toolbar.ts:84]:
```
    const iconImage = new Image();
    iconImage.className = Toolbar.classnames.iconImg;
    iconImage.src = backgroundImage;
```

It's used like this so it is just an URI. It _can_ be a base64 encoded image, or it can be an URI. So,
`backgroundImageURI` is probably the right name (i.e. not URL).

However, I'm wondering if it's correct to enforce this - there might be cases where users would want to provide icons in another way, e.g. a SVG. Not sure how to deal with this, but we should think about it

File: viewer/webpack.config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (larsmoa) [reply viewer/webpack.config.js:80]: Ok, I think maybe that's what we want for the other types - can you check if we even use `file-loader` and remove it if not?

File: viewer/packages/tools/src/Toolbar/DefaultToolbar.test.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  });
+
+  test('Test default toolbar elements', () => {
+    defaultToolbar = new DefaultToolbar(viewer, model);
+
+    const toolbarContainer = canvasContainer.getElementsByClassName('reveal-viewer-toolbar-container');
+
+    expect(toolbarContainer[0].hasChildNodes).toBeTrue;
+
+    expect(toolbarContainer[0].childElementCount).toBe(5);
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply viewer/packages/tools/src/Toolbar/DefaultToolbar.test.ts:89]: The default toolbar has 5 icon button elements

File: viewer/packages/tools/src/Toolbar/DefaultToolbar.test.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    expect(toolbar).not.toBeNull();
+
+    toolbar.addToolbarItem('Tooltip1', '', true, iconClicked);
+    toolbar.addToolbarItem('Tooltip2', '', false, iconClicked);
+
+    const toolbarContainer = canvasContainer.getElementsByClassName('reveal-viewer-toolbar-container');
+
+    expect(toolbarContainer[0].hasChildNodes).toBeTrue;
+
+    expect(toolbarContainer[0].childElementCount).toBe(7);
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply viewer/packages/tools/src/Toolbar/DefaultToolbar.test.ts:106]: Default toolbar has 5 icon element & we have added 2 more into the toolbar, so totally 7

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    } else {
+      this.axisTool.dispose();
+      this.axisTool = undefined;
+    }
+  };
+
+  /**
+   * Save the screenshot of the canvas been rendered
+   */
+  private readonly screenShot = async (): Promise<void> => {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:109]: Replace all "lambda style functions" with "plain functions", e.g. 
`private readonly screenShot = async (): Promise<void> => {`
to
 `private async screenShot(): Promise<void> {`

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+import { Cognite3DViewerToolBase } from '../Cognite3DViewerToolBase';
+import { Toolbar, ToolbarPosition } from './Toolbar';
+import { AxisViewTool } from '../AxisView/AxisViewTool';
+import { CameraControlsOptions } from '@reveal/core';
+
+import iconSet12 from './icons/Cognite_Icon_Set-12.png';
+import iconSet14 from './icons/Cognite_Icon_Set-14.png';
+import iconSet19 from './icons/Cognite_Icon_Set-19.png';
+import iconSet30 from './icons/Cognite_Icon_Set-30.png';
+import iconSet32 from './icons/Cognite_Icon_Set-32.png';
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:15]: Can we rename these to something more descriptive?

File: documentation/docs/examples/toolbar.mdx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+Reveal provides feature for using toolbar to add custom elements. Default elements for toolbar are readily available for the users to enable or disable Reveal features.
+Custom toolbar helps the user to add button to the toolbar container or to the default toolbar container.
+
+<DemoWrapper name="Cognite3DViewerDemo" />
+
+## Default Toolbar elements
+Default toolbar elements are listed below which can be used to activate or deactivate features in Reveal.
+- Take Screenshot - `addAxisToolToggle()`.
+- Enable/Disable Axis Tool - `addTakeScreenshotTool()`.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:17]: Wrong ;)

File: documentation/docs/examples/toolbar.mdx [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+- Enable/Disable Camera Target on Click - `addCameraTargetOnClickToggle()`.
+- Toggle Camera Zoom-past/Zoom-to Cursor - `addZoomPastToCursorToggle()`.
+- Fit Camera to Model - `addFitCameraToModel()`.
+
+To create and use default elements in toolbar.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:22]:
```suggestion
These can be added to the toolbar on construction:
```

File: documentation/docs/examples/toolbar.mdx [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+const toolbar = new ToolbarTool(viewer);
+
+const callbackMsg = () => (
+    alert("result called")
+);
+
+toolbar.addToolbarItem('Explode View', iconSet42, true, callbackMsg);
+toolbar.addToolbarItem('Maps', iconSet54, false, callbackMsg);
+toolbar.addToolbarItem('Settings', iconSet63, true, callbackMsg);
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:61]: Replace the random names with better names to make the example clearer

File: examples/src/pages/Toolbar.tsx [hunk 14]
```diff
@@ -0,0 +1,242 @@
+/*
```
Comment (larsmoa) [inline examples/src/pages/Toolbar.tsx:1]: I think we should just extend the Migration-example to also have a toolbar :)

File: viewer/packages/tools/src/Toolbar/ToolbarTool.test.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+import { Cognite3DViewer, CogniteModelBase } from '@reveal/core';
+import { CogniteClient } from '@cognite/sdk';
+import { ToolbarTool } from './ToolbarTool';
+import { createGlContext, mockClientAuthentication } from '../../../../test-utilities';
+import * as THREE from 'three';
+
+describe('ToolbarTool', () => {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.test.ts:10]:
```suggestion
describe(ToolbarTool.name, () => {
```

File: viewer/packages/tools/src/Toolbar/ToolbarTool.test.ts [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+    const context = createGlContext(64, 64, { preserveDrawingBuffer: true });
+    const renderer = new THREE.WebGLRenderer({ context });
+    canvasContainer = document.createElement('div');
+    canvasContainer.style.width = '640px';
+    canvasContainer.style.height = '480px';
+
+    viewer = new Cognite3DViewer({ domElement: canvasContainer, sdk, renderer });
+    models = viewer.models.slice();
+    tool = new ToolbarTool(viewer, models[0]);
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.test.ts:34]: How does this work? `viewer.models` will always be empty and `models[0]` would be undefined?

Comment (larsmoa) [reply viewer/packages/tools/src/Toolbar/ToolbarTool.test.ts:34]: See comment below on ToolbarTool and that it shouldnt be bound to a single model

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+ * The tool also helps in adding or create custom toolbar.
+ */
+export class ToolbarTool extends Cognite3DViewerToolBase {
+  private readonly _viewer: Cognite3DViewer;
+  private readonly _model: CogniteModelBase;
+  private readonly _toolbar: Toolbar;
+  private axisTool: AxisViewTool | undefined;
+  private cameraControlOption: CameraControlsOptions;
+
+  constructor(viewer: Cognite3DViewer, model: CogniteModelBase) {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:28]: Avoid limiting `ToolbarTool` to only work on a single model. I think the only function that requires model is `fitCameraToModel`. I think we instead should add a `Cognite3dViewer.fitCameraToAllModels` and use this from the tool.

Comment (pramodcog) [reply viewer/packages/tools/src/Toolbar/ToolbarTool.ts:28]: `Cognite3dViewer.fitCameraToAllModels` and use it from the tool make sense, will add and check the result

File: examples/src/pages/Toolbar.tsx (hunk 14 above)
Comment (pramodcog) [reply examples/src/pages/Toolbar.tsx]: Merged into Migration-example

File: documentation/docs/examples/toolbar.mdx [hunk 18]
```diff
@@ -0,0 +1,76 @@
+---
+id: toolbartool
+title: Toolbar tool
+describe: Toolbar container with Default Toolbar and add custom toolbar items to access features within Reveal. 
+---
+
+import { DemoWrapper } from '@site/docs/components/DemoWrapper';
+
+Reveal provides feature for using toolbar to add custom elements. Default elements for toolbar are readily available for the users to enable or disable Reveal features.
+Custom toolbar helps the user to add button to the toolbar container or to the default toolbar container.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:10]:
```suggestion
Reveal provides a toolbar to add custom elements. Default elements for toolbar are readily available for the users to enable or disable Reveal features. The toolbar can be customized to help the user to add custom buttons.
```

File: documentation/docs/examples/toolbar.mdx [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+<DemoWrapper name="Cognite3DViewerDemo" />
+
+## Default Toolbar elements
+Default toolbar elements are listed below which can be used to activate or deactivate features in Reveal.
+- Take Screenshot - `addTakeScreenshotTool()`.
+- Enable/Disable Axis Tool - `addAxisToolToggle()`.
+- Enable/Disable Camera Target on Click - `addCameraTargetOnClickToggle()`.
+- Toggle Camera Zoom-past/Zoom-to Cursor - `addZoomPastToCursorToggle()`.
+- Fit Camera to Model - `addFitCameraToModel()`.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:20]:
```suggestion
## Default toolbar elements
Supported default toolbar elements are listed below which can be used to activate or deactivate features in Reveal, or perform common actions.
- Take screenshot - `addTakeScreenshotTool()`.
- Enable/Disable axis tool - `addAxisToolToggle()`.
- Enable/Disable camera target on click - `addCameraTargetOnClickToggle()`.
- Toggle camera zoom-past/zoom-to Cursor mode - `addZoomPastToCursorToggle()`.
- Fit camera to model - `addFitCameraToModel()`.
```

File: documentation/docs/examples/toolbar.mdx [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+toolbar.addAxisToolToggle();
+toolbar.addTakeScreenshotTool();
+toolbar.addCameraTargetOnClickToggle();
+toolbar.addZoomPastToCursorToggle();
+toolbar.addFitCameraToModel();
+```
+
+## Add elements
+Each toolbar element in Reveal provides flexibilty for the user to add tooltip, icon image and a callback function to trigger when the icon button is clicked.
+We add a new element to the toolbar with `addToolbarItem(toolTip: string, backgroundImageUri: string, isToggle: boolean, onClick: () => void)`
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:37]:
```suggestion
New elements can be added to the toolbar using `addToolbarItem(toolTip: string, backgroundImageUri: string, isToggle: boolean, onClick: () => void)`
```

File: documentation/docs/examples/toolbar.mdx [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+const callbackMsg = () => (
+    alert("result called")
+);
+
+toolbar.addToolbarItem('Explode View', explodeIcon, true, callbackMsg);
+toolbar.addToolbarItem('Maps', geoMapicon, false, callbackMsg);
+toolbar.addToolbarItem('Clipping', clippingIcon, true, callbackMsg);
+```
+
+## Toolbar Position
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:64]:
```suggestion
## Toolbar position
```

File: documentation/docs/examples/toolbar.mdx [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+- `top`
+- `bottom`
+- `left`
+- `right`
+
+```jsx
+toolbar.setPosition('top');
+```
+:::note
+Default position of toolbar is at the Bottom of the window.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:75]:
```suggestion
Default position of toolbar is at the bottom of the window.
```

File: viewer/packages/tools/src/Toolbar/Toolbar.ts [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  /**
+   * Create & adds a button icons into the toolbar container
+   * @param toolTip Tooltip message to added for the icon
+   * @param backgroundImageUri Icon image to be displayed
+   * @param isToggle Is the icon button used as toggle
+   * @param onClick Click event action for the icon
+   */
+  public addToolbarItem(toolTip: string, backgroundImageUri: string, isToggle: boolean, onClick: () => void): void {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/Toolbar.ts:78]: I think it would make sense to have `addToolbarButton` and `addToolbarToogleButton`. The signature for `addToolbarToogleButton` can be e.g.
```
addToolbarToogleButton(backgroundImageUri: string, isToggle: boolean, onToggled: (isToggled: boolean) => void, toolTip: string = '');
```

Also, add tooltip as the last argument and make it optional.

**Review by larsmoa — CHANGES_REQUESTED** [review]
See comments.

Also, can we try to adopt this look:
image

Not sure where these icons are from, but I know what app they are from so contact me and I can find a name of someone to talk to.

**Comment by larsmoa** [comment]
> 

I guess we can copy them?

**Comment by larsmoa** [comment]
> > See comments.
> > Also, can we try to adopt this look: image
> > Not sure where these icons are from, but I know what app they are from so contact me and I can find a name of someone to talk to.
> 
> I got to know that these images/icon are from https://github.com/cognitedata/cogs.js. Need to check them it out if it can be integrated in Reveal or else how we can utilize part of it (like only the images for icons)

I guess we can just copy them @pramodcog ?

**Comment by pramodcog** [comment]
> > > See comments.
> > > Also, can we try to adopt this look: image
> > > Not sure where these icons are from, but I know what app they are from so contact me and I can find a name of someone to talk to.
> > 
> > 
> > I got to know that these images/icon are from https://github.com/cognitedata/cogs.js. Need to check them it out if it can be integrated in Reveal or else how we can utilize part of it (like only the images for icons)
> 
> I guess we can just copy them @pramodcog ?

I feel that copy only the css part as everything cogs.js is in React will work for us

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+   * Enable or Disable Axis view tool
+   */
+  private axisView(): void {
+    if (this.axisTool === undefined) {
+      this.axisTool = new AxisViewTool(this._viewer);
+    } else {
+      this.axisTool.dispose();
+      this.axisTool = undefined;
+    }
+  }
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:125]: Rename - it's not clear what `axisview()` does.

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+   * Save the screenshot of the canvas been rendered
+   */
+  private async screenShot(): Promise<void> {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:130]: Rename, e.g. `saveScreenshot()`

**Review by larsmoa — CHANGES_REQUESTED** [review]
See comments. Also, run the SVGs through a compressor, e.g. https://vecta.io/nano

**Comment by larsmoa** [comment]
Clicks on the toolbar is passed through the toolbar and causes a click even to be triggered on the 3D model

**Comment by larsmoa** [comment]
Replace 
- Fit camera to models (image) changed to image
- Take screenshots (image) changed to image
- Change target on click (image) changed to 
image

Having a hard time with 
image. Maybe a combination of 
image
 and 
image
or something?

Might need to get some help from design on icons...

File: viewer/webpack.config.js [hunk 26]
```diff
@@ -70,9 +71,13 @@ module.exports = env => {
           exclude: [/src\/.*\.test\.tsx?/, /src\/.*\/stubs\//]
         },
         {
-          test: /\.(png|svg|jpg|jpeg|gif)$/,
+          test: /\.(jpg|jpeg|gif)$/,
           use: ['file-loader']
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply viewer/webpack.config.js:75]: jpg, jepg & gif are needed for Bingmap in Maps

**Comment by christjt** [comment]
What is the status on this, @pramodcog? Are you awaiting review on something?

**Comment by pramodcog** [comment]
> What is the status on this, @pramodcog? Are you awaiting review on something?

I have raised a request you design team for Icon creation for CameraMoveToTargetOnClick & CameraZoomToCursor, waiting for their help, else I think it does not require any huge changes if any.

**Review by larsmoa — CHANGES_REQUESTED** [review]
- "Enable/Disable target on click" has no effect
- "Toggle zoom to cursor" has no effect
- The two option above needs to take into considered what the state is when the toolbar is created and not assume that it's in a given state (not sure how that works currently)
- Any idea why yarn.lock has 13000 lines changed? Can this be rolled back and re-generated to avoid the number of changed lines? No biggie, but a bit annoying.
-

**Comment by pramodcog** [comment]
> * "Enable/Disable target on click" has no effect
> * "Toggle zoom to cursor" has no effect
> * The two option above needs to take into considered what the state is when the toolbar is created and not assume that it's in a given state (not sure how that works currently)
> * Any idea why yarn.lock has 13000 lines changed? Can this be rolled back and re-generated to avoid the number of changed lines? No biggie, but a bit annoying.

- The Toggle's is working when use it. The above two option is retrieved from the current camera state and is not assumed with respect to the state of the camera.
- yarn.lock difference contains only to changes, that is, url-loader & reverted back copy-webpack-plugin, but will see if something can be done to it

File: documentation/docs/examples/toolbar.mdx [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+toolbar.addCameraTargetOnClickToggle();
+toolbar.addZoomPastToCursorToggle();
+toolbar.addFitCameraToModel();
+```
+
+## Add elements
+Each toolbar element in Reveal provides flexibilty for the user to add icon image, a callback function to trigger when the icon button is clicked and optional tooltip.
+New elements can be added to the toolbar by methods using
+`addToolbarButton(backgroundImageUri: string, onClick: () => void, toolTip: string = '')`
+- `backgroundImageUri` - Background image to be placed onto icon button. Pass the data URL of the image.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:38]:
```suggestion
- `backgroundImageUri` - Background image to be placed on the button. Can be a data URL or a "normal" URL.
```
Supports regular URLs too, right?

File: documentation/docs/examples/toolbar.mdx [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+- `onClick` - Click event callback function which will be used to perform custom functionlity provided by the user.
+- `toolTip` - Optional tooltip for the Icon button
+
+`addToolbarToggleButton(backgroundImageUri: string, onToggled: () => void, isActive: boolean, toolTip: string = '')`
+- `backgroundImageUri` - Background image to be placed onto icon button. Pass the data URL of the image.
```
Comment (larsmoa) [inline documentation/docs/examples/toolbar.mdx:43]: Change accordingly to the above suggestion.

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+      this._cameraOptions[0],
+      'Toggle Zoom past/Zoom to Cursor'
+    );
+  }
+
+  /**
+   * Adds Fit Camera to Model into Toolbar
+   */
+  public addFitCameraToModel(): void {
+    this.addToolbarButton(svgFitToCameraIcon, this._handleFitCameraToModelListener, 'Fit Camera to Model');
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:125]:
```suggestion
    this.addToolbarButton(svgFitToCameraIcon, this._handleFitCameraToModelListener, 'Fit Camera to Model');
```
Generally, don't use capital first letters. We wouldn't do this in plain english so we shouldn't in customer facing messages either.

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+import svgFitToCameraIcon from './icons/fittocamera.svg';
+
+/**
+ * Tool to help user to use the default toolbar items such as Camera & other features in Reveal to enable/disable it.
+ * The tool also helps in adding or create custom toolbar.
+ */
+export class ToolbarTool extends Cognite3DViewerToolBase {
+  private readonly _viewer: Cognite3DViewer;
+  private readonly _toolbar: Toolbar;
+  private axisTool: AxisViewTool | undefined;
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:24]:
```suggestion
  private _axisTool: AxisViewTool | undefined;
```

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+  private readonly _cameraOptions: boolean[] = [false, false];
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:25]: There should not be any reason to have this state at all - it can be determined from the camera manger. Storing state should almost always be avoided when possible.

Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:25]: Change to a struct with named options rather than using tuple.

File: viewer/packages/tools/src/Toolbar/ToolbarTool.ts [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+   * Add Save the screenshot of the canvas been rendered to the Toolbar
+   */
+  public addTakeScreenshotTool(): void {
+    this.addToolbarButton(svgScreenshotIcon, this._handleScreenshotListener, 'Take Screenshot');
+  }
+
+  /**
+   * Adds Toggle enabling/disabling the Camera target on mouse click to the Toolbar
+   */
+  public addCameraTargetOnClickToggle(): void {
```
Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:100]: This function assumes that change target initally. Use state from camera manager.

Comment (larsmoa) [inline viewer/packages/tools/src/Toolbar/ToolbarTool.ts:100]: This function assumes that change target initally. Use state from camera manager.

**Review by larsmoa — CHANGES_REQUESTED** [review]

**Comment by pramodcog** [comment]
Will revisit when necessary
